File: backend/app/ai/schemas/question_design.py
from pydantic import BaseModel, Field, field_validator


def _null_to_str(value: object) -> object:
    return "" if value is None else value


def _coerce_int(value: object, default: int) -> int:
    if value is None or value == "":
        return default
    try:
        return int(value)
    except (TypeError, ValueError):
        return default


class RevisionSuggestion(BaseModel):
    field: str = Field(description="prompt | modelAnswer | points | instructions")
    current_excerpt: str = Field(alias="currentExcerpt", default="")
    proposed_text: str = Field(alias="proposedText")
    reason: str

    model_config = {"populate_by_name": True}


class QuestionValidityItem(BaseModel):
    question_order: int = Field(alias="questionOrder")
    matched_type_label: str = Field(alias="matchedTypeLabel", default="")
    coverage: str = Field(description="sufficient | partial | insufficient")
    summary: str
    improvements: list[str] = Field(default_factory=list)
    referenced_past_questions: list[str] = Field(alias="referencedPastQuestions", default_factory=list)
    revision_suggestions: list[RevisionSuggestion] = Field(alias="revisionSuggestions", default_factory=list)

    model_config = {"populate_by_name": True}


class ValidityCheckResponse(BaseModel):
    overall_summary: str = Field(alias="overallSummary")
    university_slug: str = Field(alias="universitySlug")
    items: list[QuestionValidityItem] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class GeneratedQuestionItem(BaseModel):
    type_label: str = Field(alias="typeLabel")
    major_order: int = Field(alias="majorOrder")
    part_label: str | None = Field(alias="partLabel", default=None)
    prompt: str
    model_answer: str = Field(alias="modelAnswer")
    points: int = 10
    type: str = "english"
    answer_format: str | None = Field(alias="answerFormat", default=None)
    notes: str = ""
    reference_examples: list[str] = Field(alias="referenceExamples", default_factory=list)
    # 生徒がしがちな想定誤答（指導・採点準備用）
    anticipated_mistakes: list[str] = Field(
        alias="anticipatedMistakes", default_factory=list
    )

    model_config = {"populate_by_name": True}

    @field_validator("type_label", "prompt", "model_answer", "notes", mode="before")
    @classmethod
    def coerce_optional_strings(cls, value: object) -> object:
        return _null_to_str(value)

    @field_validator("major_order", mode="before")
    @classmethod
    def coerce_major_order(cls, value: object) -> int:
        return _coerce_int(value, 1)

    @field_validator("points", mode="before")
    @classmethod
    def coerce_points(cls, value: object) -> int:
        return _coerce_int(value, 10)

    @field_validator("part_label", mode="before")
    @classmethod
    def coerce_part_label(cls, value: object) -> object:
        if value is None or value == "":
            return None
        return str(value).strip() or None

    @field_validator("reference_examples", "anticipated_mistakes", mode="before")
    @classmethod
    def coerce_str_list(cls, value: object) -> list[str]:
        if value is None:
            return []
        if isinstance(value, str):
            return [value] if value.strip() else []
        if isinstance(value, list):
            return [str(x).strip() for x in value if str(x).strip()]
        return []


class GenerateQuestionsResponse(BaseModel):
    batch_id: str = Field(alias="batchId")
    draft_ids: list[str] = Field(alias="draftIds")
    questions: list[GeneratedQuestionItem] = Field(default_factory=list)

    model_config = {"populate_by_name": True}

File: frontend/src/types/question-design.ts
export type CoverageLevel = "sufficient" | "partial" | "insufficient";

export interface RevisionSuggestion {
  field: string;
  currentExcerpt: string;
  proposedText: string;
  reason: string;
}

export interface QuestionValidityItem {
  questionOrder: number;
  matchedTypeLabel: string;
  coverage: CoverageLevel;
  coverageLabel?: string;
  summary: string;
  improvements: string[];
  referencedPastQuestions: string[];
  revisionSuggestions: RevisionSuggestion[];
}

export interface ValidityReport {
  overallSummary: string;
  universitySlug: string;
  items: QuestionValidityItem[];
  checkedAt?: string;
}

export interface QuestionTypeCatalogItem {
  majorOrder: number;
  partLabel?: string | null;
  typeLabel: string;
  years: number[];
  sampleQuestionIds: string[];
}

/** UI 用の生成単位（第5問は1枠に集約など） */
export interface GenerationUnit {
  majorOrder: number;
  partLabel?: string | null;
  typeLabel: string;
  unitKey: string;
  years: number[];
  sampleQuestionIds: string[];
  catalogKeys: string[];
  pipeline: "q5" | "q4a" | "q4b" | "q1a" | "q1b" | "q2a" | "q2b" | "generic";
}

export interface Q5GenerationArtifacts {
  passage?: string;
  passageTitle?: string;
  fullTranslationJa?: string;
  evaluatorPassed?: boolean;
  evaluatorIssues?: string[];
  evaluatorSummary?: string;
  retriedQuestions?: boolean;
  retriedProblem?: boolean;
  themeSummary?: string;
  layout?: string;
  items?: unknown[];
}

export interface GeneratedQuestionDraft {
  id?: string;
  batchId?: string;
  teacherId?: string;
  universitySlug: string;
  typeLabel: string;
  majorOrder: number;
  partLabel?: string | null;
  prompt: string;
  modelAnswer: string;
  points: number;
  type: string;
  answerFormat?: string | null;
  notes?: string;
  referenceExamples?: string[];
  /** 生徒がしがちな想定誤答（採点・指導の準備用） */
  anticipatedMistakes?: string[];
  status?: "draft" | "promoted";
  difficulty?: string;
  topicHint?: string;
  createdAt?: string;
  generationPipeline?: "q5" | string;
  generationArtifacts?: Q5GenerationArtifacts;
}

export interface GenerateQuestionsResponse {
  batchId: string;
  draftIds: string[];
  questions: GeneratedQuestionDraft[];
}

/** ② セット下書き（複数問＋自動検証込み） */
export interface TestDraftSet {
  id: string;
  teacherId?: string;
  title: string;
  universitySlug: string;
  studentId?: string;
  studentName?: string;
  /** カルテ由来の弱点フォーカス（生徒指定時のみ） */
  weaknessFocus?: string;
  difficulty?: string;
  topicHint?: string;
  referenceYears?: number[];
  status?: "draft" | "promoted";
  reviewStatus?: "draft" | "confirmed";
  questions: GeneratedQuestionDraft[];
  questionCount: number;
  totalPoints: number;
  /** 自動で実行した妥当性検証の結果（過去問が無い場合は null） */
  validityReport?: ValidityReport | null;
  /** 検証で不足が出たため1回再生成したか */
  autoRetried?: boolean;
  createdAt?: string;
}

export interface BuildTestDraftBody {
  selections: Array<{ majorOrder: number; partLabel?: string | null; typeLabel?: string }>;
  referenceYears?: number[];
  difficulty?: string;
  topicHint?: string;
  countPerType?: number;
  studentId?: string;
  title?: string;
}
